### PR TITLE
support discord

### PR DESCRIPTION
**Pull request by Gentleelephant**
Signed-off-by: Gentleelephant <birdhk@kubesphere.io>

### CODE REVIEW

File: pkg/apis/v2beta2/receiver_types.go [hunk 1]
```diff
@@ -207,6 +207,33 @@ type WechatChatBot struct {
...
+	Type *string `json:"type,omitempty"`
+
+	// Mentioned users
+	MentionedUsers []string `json:"mentionedUsers,omitempty"`
+
+	// Mentioned roles
+	MentionedRoles []string `json:"mentionedRoles,omitempty"`
+
+	// DiscordConfigSelector to be selected for this receiver
+	DiscordConfigSelector *metav1.LabelSelector `json:"webhookConfigSelector,omitempty"`
```
Comment (wanjunlei) [inline pkg/apis/v2beta2/receiver_types.go:231]: The `discord` has no config, no need for this.

File: pkg/internal/discord/types.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	"strconv"
+)
+
+type Receiver struct {
+	*internal.Common
+	Webhook        *v2beta2.Credential `json:"webhook"`
+	Type           *string             `json:"type,omitempty"`
+	MentionedUsers []string            `json:"mentionedUsers,omitempty"`
+	MentionedRoles []string            `json:"mentionedRoles,omitempty"`
+	*Config
```
Comment (wanjunlei) [inline pkg/internal/discord/types.go:18]: No need for this.

File: pkg/internal/discord/types.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+		MentionedRoles: discord.MentionedRoles,
+		MentionedUsers: discord.MentionedUsers,
+		Type:           discord.Type,
+	}
+
+	if discord.Webhook != nil {
+		r.Webhook = discord.Webhook
+	}
+
+	r.ResourceVersion, _ = strconv.ParseUint(obj.ResourceVersion, 10, 64)
```
Comment (wanjunlei) [inline pkg/internal/discord/types.go:49]: What is the purpose of `ResourceVersion`?

File: pkg/notify/notifier/discord/discord.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+		}
+	}
+
+	n.receiver = receiver.(*discord.Receiver)
+	if utils.StringIsNil(n.receiver.TmplType) {
+		n.receiver.TmplType = tmplType
+	}
+
+	if utils.StringIsNil(n.receiver.TmplName) {
+		if tmplName != "" {
```
Comment (wanjunlei) [inline pkg/notify/notifier/discord/discord.go:76]: Can the `content` and `embed` messages use the same template?

File: pkg/notify/notifier/discord/discord.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+		request.Header.Set("Content-Type", "application/json; charset=utf-8")
+		client := http.Client{}
+		resp, err := client.Do(request)
+		if err != nil {
+			return true, err
+		}
+		code := resp.StatusCode
+		level.Debug(n.logger).Log("msg", "DiscordNotifier", "response code:", code)
+
+		if code != http.StatusNoContent {
```
Comment (wanjunlei) [inline pkg/notify/notifier/discord/discord.go:170]: Does the discord server return a 204 for sending successfully, not 200?

File: pkg/notify/notifier/discord/discord.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+		return false, nil
+	}
+
+	start := time.Now()
+	defer func() {
+		_ = level.Debug(n.logger).Log("msg", "DiscordNotifier: send message", "used", time.Since(start).String())
+	}()
+
+	retry := 0
+	for {
```
Comment (wanjunlei) [inline pkg/notify/notifier/discord/discord.go:182]: Will it retry infinitely under extreme conditions?

File: pkg/notify/notifier/discord/discord.go (hunk 4 above)
Comment (Gentleelephant) [reply pkg/notify/notifier/discord/discord.go:76]: yes,they only differ in the display.
content:
image
embed:
image

File: pkg/notify/notifier/discord/discord.go (hunk 5 above)
Comment (Gentleelephant) [reply pkg/notify/notifier/discord/discord.go:170]: yes
image

File: pkg/notify/notifier/discord/discord.go (hunk 6 above)
Comment (Gentleelephant) [reply pkg/notify/notifier/discord/discord.go:182]: Should we limit the number of times？

Comment (wanjunlei) [reply pkg/notify/notifier/discord/discord.go:183]: Yes

File: pkg/internal/discord/types.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+	if discord.Webhook != nil {
+		r.Webhook = discord.Webhook
+	}
+
+	return r
+}
+
+func (r *Receiver) SetConfig(c internal.Config) {
+	if reflect2.IsNil(c) {
```
Comment (wanjunlei) [inline pkg/internal/discord/types.go:50]: Just return, no need to add this if statement.

File: pkg/internal/discord/types.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+func NewConfig(obj *v2beta2.Config) internal.Config {
+	return nil
+}
+
+func (c *Config) Validate() error {
+	return nil
+}
+
+func (c *Config) Clone() internal.Config {
+	return &Config{
```
Comment (wanjunlei) [inline pkg/internal/discord/types.go:90]: Just return.

File: pkg/notify/notifier/discord/discord.go (hunk 6 above)
Comment (wanjunlei) [reply pkg/notify/notifier/discord/discord.go:183]: 120 is too large, I think 3 is more suitable.

**Review by wanjunlei — APPROVED** [review]

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**